internal/resource/config/datastore/data_store_resource.go

```diff
@@ -0,0 +1,290 @@
+	schema := schema.Schema{
+		Description: "Manages a Data Store",
+		Attributes: map[string]schema.Attribute{
+			"mask_attribute_values": schema.BoolAttribute{
+				Description: "Whether attribute values should be masked in the log.",
+				Computed:    true,
+				Optional:    true,
+				Default:     booldefault.StaticBool(false),
+			},
+			"custom_data_store":                toSchemaCustomDataStore(),
+			"jdbc_data_store":                  toSchemaJdbcDataStore(),
+			"ldap_data_store":                  toSchemaLdapDataStore(),
+			"ping_one_ldap_gateway_data_store": toSchemaPingOneLdapGatewayDataStore(),
+		},
+	}
+	id.ToSchema(&schema)
+	id.ToSchemaCustomId(&schema, false,
+		"The persistent, unique ID for the data store. It can be any combination of [a-zA-Z0-9._-]. This property is system-assigned if not specified.")
+
+	resp.Schema = schema
+}
+
+func (r *dataStoreResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_data_store"
+}
+
+func (r *dataStoreResource) Configure(_ context.Context, req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	providerCfg := req.ProviderData.(internaltypes.ResourceConfiguration)
+	r.providerConfig = providerCfg.ProviderConfig
+	r.apiClient = providerCfg.ApiClient
+
+}
+
+func (r *dataStoreResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	var plan *dataStoreResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	var respDiags diag.Diagnostics
+
+	if plan == nil {
+		return
+	}
+
+	// Build name attribute for data stores that have it
+	if internaltypes.IsDefined(plan.JdbcDataStore) {
+		jdbcDataStore := plan.JdbcDataStore.Attributes()
+		if !internaltypes.IsDefined(jdbcDataStore["name"]) {
+			namePrefix := func() string {
+				topConnectionUrl := jdbcDataStore["connection_url"].(types.String)
+				if internaltypes.IsNonEmptyString(topConnectionUrl) {
+					return topConnectionUrl.ValueString()
+				}
+				connectionUrlTags := jdbcDataStore["connection_url_tags"].(types.Set)
+				if len(connectionUrlTags.Elements()) > 0 {
+					connectionUrlTagsFirstElem := connectionUrlTags.Elements()[0].(types.Object).Attributes()
+					return connectionUrlTagsFirstElem["connection_url"].(types.String).ValueString()
+				}
+				return ""
+			}
+			userName := jdbcDataStore["user_name"].(types.String).ValueString()
+			jdbcDataStore["name"] = types.StringValue(namePrefix() + " (" + userName + ")")
+			plan.JdbcDataStore, respDiags = types.ObjectValue(jdbcDataStoreAttrType, jdbcDataStore)
+			resp.Diagnostics.Append(respDiags...)
+		}
+	}
+
+	if internaltypes.IsDefined(plan.LdapDataStore) {
+		ldapDataStore := plan.LdapDataStore.Attributes()
+		if !internaltypes.IsDefined(ldapDataStore["name"]) {
```

Comment:
Rather than using IsDefined for setting defaults, we should use IsUnknown. We don't need to overwrite nulls, just unknowns

Reply:
changed to use IsUnknown()